File: schemas/media-streaming/convex/play_history.ts
// play_history: per-user track listening history with context information.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const play_history = defineTable({
  userId: v.id("users"),
  trackId: v.id("tracks"),
  durationMs: v.number(),
  completed: v.boolean(),
  contextType: v.union(
    v.literal("album"),
    v.literal("playlist"),
    v.literal("artist"),
    v.literal("chart"),
    v.literal("search"),
    v.literal("queue"),
    v.literal("unknown")
  ),
  contextId: v.optional(v.string()),
  playedAt: v.number(),
})
  .index("by_user_id_played_at", ["userId", "playedAt"])
  .index("by_track_id_played_at", ["trackId", "playedAt"]);

File: schemas/media-streaming/spacetimedb/play_history.rs
// play_history: User listening history with playback context and completion tracking.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum PlayContextType {
    Album, // type: String
    Playlist,
    Artist,
    Chart,
    Search,
    Queue,
    Unknown,
}

#[spacetimedb::table(name = play_history, public)]
pub struct PlayHistory {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub user_id: String, // UUID — FK → users.id (cascade delete)
    #[index(btree)]
    pub track_id: String, // UUID — FK → tracks.id (cascade delete)
    pub duration_ms: i32,
    pub completed: bool,
    pub context_type: PlayContextType,
    pub context_id: Option<String>,
    pub played_at: Timestamp,
}
// Composite index: (user_id, played_at) — not supported; btree on user_id covers leading column.
// Composite index: (track_id, played_at) — not supported; btree on track_id covers leading column.
